### PR TITLE
optimize stride size for update FieldJ

### PR DESCRIPTION
**Pull request by psychocoderHPC**
The stride size for `StrideMapping` depends on the supercell size and the stencil margins.
In the case where the lower and upper margin fits into one supercell we need only a stride of 2 instead of the current used default stride 3.
If we use stencils with margins larger than one supercell the default 3 can be to low and will produce data race conditions (therefore this PR is also a fix for big stencil margins).

- calculate optimal stride size (skip only as much supercells as needed to avoid race conditions for cell updates)

- [x] depends on #2612

### CODE REVIEW

File: include/picongpu/fields/FieldJ.tpp [hunk 1]
```diff
@@ -278,7 +278,24 @@ void FieldJ::computeCurrent( ParticlesClass &parClass, uint32_t )
...
+     * between different blocks in the kernel
+     */
+    constexpr uint32_t skipSuperCells = (MaxMargin::value + SuperCellMinSize::value - 1) / SuperCellMinSize::value;
+    StrideMapping<
+        AREA,
+        skipSuperCells + 1, // stride 1 means each supercell is used
+        MappingDesc
+    > mapper( cellDescription );
+
+    std::cout<<skipSuperCells<<std::endl;
```
Comment (ax3l) [inline include/picongpu/fields/FieldJ.tpp]: leftover debug :)

File: include/picongpu/fields/FieldJ.tpp [hunk 2]
```diff
@@ -278,7 +278,24 @@ void FieldJ::computeCurrent( ParticlesClass &parClass, uint32_t )
...
+    using MaxMargin = typename pmacc::math::CT::max< MarginPerDim >::type;
+    using SuperCellMinSize = typename pmacc::math::CT::min< SuperCellSize >::type;
+
+    /* number of supercells which must be skipped to avoid overlapping areas
+     * between different blocks in the kernel
+     */
+    constexpr uint32_t skipSuperCells = (MaxMargin::value + SuperCellMinSize::value - 1) / SuperCellMinSize::value;
+    StrideMapping<
+        AREA,
+        skipSuperCells + 1, // stride 1 means each supercell is used
```
Comment (ax3l) [inline include/picongpu/fields/FieldJ.tpp]: 1u

**Comment by psychocoderHPC** [comment]
ready to merge

File: include/picongpu/fields/FieldJ.tpp [hunk 3]
```diff
@@ -278,7 +278,23 @@ void FieldJ::computeCurrent( ParticlesClass &parClass, uint32_t )
         typename GetMargin<ParticleCurrentSolver>::UpperMargin
     > BlockArea;
 
-    StrideMapping<AREA, 3, MappingDesc> mapper( cellDescription );
+    using MarginPerDim = typename pmacc::math::CT::add<
```
Comment (ax3l) [inline include/picongpu/fields/FieldJ.tpp]: you will need to add a comment here to describe the logic (see PR description).

File: include/picongpu/fields/FieldJ.tpp [hunk 4]
```diff
@@ -278,7 +278,23 @@ void FieldJ::computeCurrent( ParticlesClass &parClass, uint32_t )
...
+        typename GetMargin<ParticleCurrentSolver>::LowerMargin,
+        typename GetMargin<ParticleCurrentSolver>::UpperMargin
+    >::type;
+    using MaxMargin = typename pmacc::math::CT::max< MarginPerDim >::type;
+    using SuperCellMinSize = typename pmacc::math::CT::min< SuperCellSize >::type;
+
+    /* number of supercells which must be skipped to avoid overlapping areas
+     * between different blocks in the kernel
+     */
+    constexpr uint32_t skipSuperCells = (MaxMargin::value + SuperCellMinSize::value - 1) / SuperCellMinSize::value;
```
Comment (ax3l) [inline include/picongpu/fields/FieldJ.tpp]: `- 1u`

**Pull request merged**